rviz_simulator/src/camera.cpp

```diff
@@ -0,0 +1,505 @@
+  s << "Feedback from camera '" << feedback->marker_name << "' "
+    << " / control '" << feedback->control_name << "'";
+
+  std::ostringstream mouse_point_ss;
+  if (feedback->mouse_point_valid)
+  {
+    mouse_point_ss << " at " << feedback->mouse_point.x << ", " << feedback->mouse_point.y << ", "
+                   << feedback->mouse_point.z << " in frame " << feedback->header.frame_id;
+  }
+
+  switch (feedback->event_type)
+  {
+    case visualization_msgs::InteractiveMarkerFeedback::BUTTON_CLICK:
+      ROS_INFO_STREAM(s.str() << ": button click, camera taking picture" << mouse_point_ss.str() << ".");
+      this->takePicture();
+      break;
+
+    case visualization_msgs::InteractiveMarkerFeedback::POSE_UPDATE:
+      ROS_INFO_STREAM(s.str() << ": pose changed"
+                              << "\nposition = " << feedback->pose.position.x << ", " << feedback->pose.position.y
+                              << ", " << feedback->pose.position.z << "\norientation = " << feedback->pose.orientation.w
+                              << ", " << feedback->pose.orientation.x << ", " << feedback->pose.orientation.y << ", "
+                              << feedback->pose.orientation.z << "\nframe: " << feedback->header.frame_id << " time: "
+                              << feedback->header.stamp.sec << "sec, " << feedback->header.stamp.nsec << " nsec");
+      visualization_msgs::InteractiveMarker temp;
+      break;
+  }
+
+  this->g_interactive_marker_server_->applyChanges();
+}
+
+Eigen::Affine3d Camera::calculateReference_T_Target(visualization_msgs::InteractiveMarker& reference,
+                                                    visualization_msgs::InteractiveMarker& target)
+{
+  Eigen::Affine3d origin_T_reference, origin_T_target;
+  tf::poseMsgToEigen(reference.pose, origin_T_reference);
+  tf::poseMsgToEigen(target.pose, origin_T_target);
+  return origin_T_reference.inverse() * origin_T_target;
+}
+
+void Camera::takePicture()
+{
+  if (!this->first_picture_taken_)
+  {
+    this->calculateWorld_T_Targets("targets.yaml");
+    this->first_picture_taken_ = true;
+  }
+
+  // YAML Emitter for the detections_X.yaml file
+  YAML::Emitter detections_out;
+  detections_out << YAML::BeginMap;  // level 0
+
+  // calculating world_T_camera and dumping to YAML::Emitter
+  visualization_msgs::InteractiveMarker camera_marker;
+  this->g_interactive_marker_server_->get(this->marker_name_, camera_marker);
+
+  visualization_msgs::InteractiveMarker world_marker;
+  this->g_interactive_marker_server_->get("tag0", world_marker);
+
+  Eigen::Affine3d world_T_camera = this->calculateReference_T_Target(world_marker, camera_marker);
+  this->Affine3dToYaml(world_T_camera, "world_T_camera", detections_out);
+
+  detections_out << YAML::Key << "detections";
+  detections_out << YAML::Value << YAML::BeginSeq;  // detections sequence
+
+  // processing all camera_T_target transforms
+  for (int i = 0; i < this->world_T_targets_.size(); i++)
```

Comment:
You can use the iterating options of a vector to iterate over the targets
either
https://en.cppreference.com/w/cpp/container/vector
```suggestion
  for (Eigen::Affine3d world_T_target : this->world_T_targets_)
```
or 
http://www.cplusplus.com/reference/vector/vector/begin/
```suggestion
  for (std::vector<Eigen::Affine3d>::iterator it =  this->world_T_targets_.begin() ; it !=  this->world_T_targets_.end(); ++it)

```